### PR TITLE
Add getters and setters for ShardChains, ExecutionEnvironments, and ShardBlocks

### PR DESCRIPTION
**Pull request by gjtrowbridge**
Also allows for the retrieval of general ethereum simulation state, and implements basic `From` or `TryFrom` methods for converting between external-facing interface arguments and their (if any) corresponding internal representations.

### CODE REVIEW

**Comment by gjtrowbridge** [comment]
I considered splitting this into smaller diffs but decided this was easier to read as a whole and still not too large of a change. Please let me know if you feel differently as you review this: I'd like to (and believe as a team we should get into the habit of) making reviews easy to read and small enough to digest easily.

File: notion-server/src/ethereum.rs [hunk 1]
```diff
@@ -1,4 +1,77 @@
...
+    },
+    OutOfBounds {
+        message: String,
+    }
+}
+
+pub mod interface_args {
+    #[derive(Debug, Default)]
+    pub struct GeneralSimulationState {
+        pub num_execution_environments: u32,
```
Comment (SamWilsn) [inline notion-server/src/ethereum.rs:22]: Since we're the only downstream crate of this library, this feedback isn't _super_ necessary here. I'm still going to take the opportunity to write up my thoughts and maybe provide some useful Rust information for the future!

Adding a new public field to a struct is a breaking change, unless that struct already has at least one private field. This follows from the initialization syntax having to specify every field in the struct.

For public types, generally I'd suggest using one of several alternatives:

### Constructor & Everything Private

Best for structs that have few members, or members that can easily be initialized to default values. The drawbacks are long function signatures, and having to define getters for everything.

```rust
pub struct Foo {
    bar: i32,
}

impl Foo {
    pub fn new(bar: i32) -> Self {
        Self { bar }
    }

    pub fn bar(&self) -> i32 {
        self.bar
    }
}

fn main() {
    let foo = Foo::new(5);
}
```

### One Private Field

Basically the same as the previous option, except that all fields are public, except for one zero-sized item. Avoids getters and setters, but either requires a long constructor or fields which can cheaply be default initialized.

```rust
struct Foo {
    _p: (),
    pub bar: i32,
}

impl Default for Foo {
    pub fn default() -> Self {
        Foo {
            _p: (),
            bar: 0,
        }
}

fn main() {
    let foo = Foo::default();
    foo.bar = 5;
}
```

### Typed Builder

The `typed-builder` crate is my personal favourite way of initializing structs, and probably the easiest to avoid breaking changes. `typed-builder` will derive a builder type for your struct that handles construction elegantly.

```rust
#[macro_use]
extern crate typed_builder;

#[derive(TypedBuilder)]
struct Foo {
    bar: i32,
}

fn main() {
    let foo = Foo::builder()
        .bar(5)
        .build();
}
```

File: notion-server/src/ethereum.rs [hunk 2]
```diff
@@ -1,4 +1,77 @@
...
+        pub num_shard_chains: u32,
+    }
+
+    #[derive(Debug, Default)]
+    pub struct ExecutionEnvironment {
+        pub base64_encoded_wasm_code: String,
```
Comment (SamWilsn) [inline notion-server/src/ethereum.rs:28]: I think I'd prefer to store this as a `Vec<u8>` even in the argument, and implement the conversion to/from base64 in the serialization code. Serde has [an attribute](https://serde.rs/field-attrs.html#with) for exactly that.

Not a blocking issue for the PR, just a suggestion.

File: notion-server/src/ethereum.rs [hunk 3]
```diff
@@ -1,4 +1,77 @@
...
+        source: base64::DecodeError,
+    },
+    OutOfBounds {
+        message: String,
+    }
+}
+
+pub mod interface_args {
+    #[derive(Debug, Default)]
+    pub struct GeneralSimulationState {
```
Comment (SamWilsn) [inline notion-server/src/ethereum.rs:21]: Unless there is a need to differentiate `GeneralSimulationState` and `SimulationState`, I think I'd prefer the latter.

File: notion-server/src/ethereum.rs [hunk 4]
```diff
@@ -1,4 +1,77 @@
...
+        pub num_execution_environments: u32,
+        pub num_shard_chains: u32,
+    }
+
+    #[derive(Debug, Default)]
+    pub struct ExecutionEnvironment {
+        pub base64_encoded_wasm_code: String,
+    }
+
+    impl From<&super::ExecutionEnvironment> for ExecutionEnvironment {
```
Comment (SamWilsn) [inline notion-server/src/ethereum.rs:31]: I love the use of Rust's standard conversion traits! That's awesome.

If it's at all possible, taking a `super::ExecutionEnvironment` instead of a reference would be arguably better. Using a reference means that every conversion from an `args::ExecutionEnvironment` to a `super::ExecutionEnvironment` will copy memory. Using the structs directly means that less allocations happen overall.

In this case, I think that because you're converting from base64 `String` to byte arrays, you'll end up with allocations anyway. If you decide to apply [this comment](#r350313031), you should probably do this one too, otherwise it's less important.

File: notion-server/src/ethereum.rs [hunk 5]
```diff
@@ -1,4 +1,77 @@
...
+            ShardBlock {
+                transactions,
+            }
+        }
+    }
+
+    #[derive(Debug, Default, Eq, PartialEq)]
+    pub struct ShardTransaction {
+        pub base64_encoded_data: String,
+        pub ee_index: u32,
```
Comment (SamWilsn) [inline notion-server/src/ethereum.rs:61]: Are we going with typed execution environment indexes (i.e. `struct EeIndex(i32)`) or no?

File: notion-server/src/ethereum.rs [hunk 6]
```diff
@@ -14,24 +87,40 @@ impl EthereumSimulation {
         }
     }
 
+    pub fn get_general_simulation_state(&self) -> interface_args::GeneralSimulationState {
```
Comment (SamWilsn) [inline notion-server/src/ethereum.rs:90]: Getters in Rust generally don't include the `get_` prefix. Perhaps `pub fn simulation_state(&self)`?

File: notion-server/src/ethereum.rs [hunk 7]
```diff
@@ -1,4 +1,77 @@
...
+    Decode {
+        backtrace: Backtrace,
+        source: base64::DecodeError,
+    },
+    OutOfBounds {
+        message: String,
+    }
+}
+
+pub mod interface_args {
```
Comment (SamWilsn) [inline notion-server/src/ethereum.rs:19]: Why the rename?

File: notion-server/src/ethereum.rs (hunk 1 above)
Comment (gjtrowbridge) [reply notion-server/src/ethereum.rs:22]: Very informative comment, thanks!  Worth mentioning that as a return value it's not necessarily a breaking change to add additional public fields, but I agree that for any incoming request (say, to the web server REST api), we'd need to create the appropriate structs based eg. on the incoming JSON data.  That said, I think the "breaking" part would depend on which interface we expect folks to use.  If we want to make this `EthereumSimulation` interface usable to all, then we'd want to have the defaults, etc handled there.  However, if the boundary is the web server REST API (or other external-facing API), then it wouldn't be a breaking change to clients of that API unless our REST server couldn't or didn't provide defaults / fallbacks to provide backward compatibility if things change in the future.  I guess my point is that everything you said is correct, but we'll always need to address the possibility of breaking changes regardless, and would need to do so wherever we want to consider the start of the public-facing interface.

That said, I like the TypedBuilder pattern, seems pretty slick, and I'm certainly on board with incorporating that into our code.  And my impression from your comment is that you think we should make `EthereumSimulation` as breaking-change-proof / backward compatible as possible.  Assuming I'm interpreting that right, I also agree with you there, though for now I think it's OK to not worry about it *too* much since as you said our web server will be the only downstream consumer, and it may save time during the prototyping phase to not worry about being breaking-change compatible right off the bat.

TL;DR: I like the TypedBuilder pattern, and think it's a great suggestion to think about using it.

File: notion-server/src/ethereum.rs (hunk 2 above)
Comment (gjtrowbridge) [reply notion-server/src/ethereum.rs:28]: I'm on board for that if we'd rather do the serialization in the web handlers.  I think that might be better / would keep the `EthereumSimulation` interface cleaner and more flexible, so sounds good to me.

File: notion-server/src/ethereum.rs (hunk 3 above)
Comment (gjtrowbridge) [reply notion-server/src/ethereum.rs:21]: Yeah I don't feel strongly here.  My thinking was that `SimulationState` could be interpreted to mean "everything the simulation state contains" vs. `GeneralSimulationState` felt more likely to be interpreted as "an broad overview of what the simulation state looks like".  Since this struct is meant to represent the latter, I chose the latter name.  Definitely open to discussing this or shortening it, though.

File: notion-server/src/ethereum.rs (hunk 4 above)
Comment (gjtrowbridge) [reply notion-server/src/ethereum.rs:31]: I think we *want* to use a reference here, though I'd be happy to discuss more and learn if maybe I'm missing something.

These are 2 different objects (the external interface representation and the internal "true" representation), so allocating / creating a new object seems like what I'd expect to happen in a From statement like this.  Seems like making it force moves would be a more restrictive API, and could be a premature optimization (ie. I don't think this simulation will ever need to worry about allocating slightly more memory than normal).  If performance is or will become a big concern, in that case maybe this is more important to do without using a reference as input.

Agreed with your main point that if we just move over variables instead of allocating new ones that this would be more performance (eg. if we decide to change the interface_args to use Vec<u8> we could just move those over vs creating from scratch).

File: notion-server/src/ethereum.rs (hunk 5 above)
Comment (gjtrowbridge) [reply notion-server/src/ethereum.rs:61]: My intent here was to keep what I considered "internal" structs / data structures completely out of the public-facing interface.  So in this case: while `EeIndex` seemed like a great way to make our code more clear and easier to read for writing internal logic, IMO the interface will be more accessible and will be easier to use if all the external-facing types are (for the most part) either 1) common types users are used to seeing (eg. u32, String, possible Vec<u8>, etc), or 2) custom types useful for defining the interface or bundling function arguments (eg. `ShardTransaction` and the other structs in `mod interface_args`. 

Thanks for the callout, definitely interested in discussing with you more, let's chat about this next time we check in.

File: notion-server/src/ethereum.rs (hunk 6 above)
Comment (gjtrowbridge) [reply notion-server/src/ethereum.rs:90]: Works for me!  I always like to learn more about Rust packages and Rust conventions / tips, thanks for including all of them and please keep them coming!

File: notion-server/src/ethereum.rs (hunk 7 above)
Comment (gjtrowbridge) [reply notion-server/src/ethereum.rs:19]: `args` felt too generic, and it make me uncomfortable to have eg. `args::ShardBlock` and `ShardBlock`, because it wasn't immediately apparent from the module name and struct names what the difference was.  IMO `interface_args` makes it MUCH more immediately clear what the differences between the structs are, and will make 1) it easier for us to talk about these 2 structs and quickly distinguish (eg. the public-facing interface at `interface_args::ShardBlock` vs the internal representation `ShardBlock`) and 2) seems like this would also be a more immediately-obvious distinction if people new to this codebase are reading it for the first time.

Interested in your opinions here also and looking forward to discussing, but that was my logic.

**Comment by gjtrowbridge** [comment]
Responded to some comments, Sam, thanks as always for the very thorough review!  Looking forward to discussing some of these more offline.

File: notion-server/src/ethereum.rs (hunk 1 above)
Comment (SamWilsn) [reply notion-server/src/ethereum.rs:22]: > Worth mentioning that as a return value it's not necessarily a breaking change to add additional public fields.

Without a private member, any downstream crate could create an instance. So even if it wasn't intentional for some binary to be able to create a `GeneralSimulationState`, it would still be possible, and therefore be a breaking change to add a public field. Since we control the one downstream crate, it isn't an issue here.

> And my impression from your comment is that you think we should make EthereumSimulation as breaking-change-proof / backward compatible as possible.

That would be the correct impression! I am a zealot for following SemVer.

File: notion-server/src/ethereum.rs (hunk 4 above)
Comment (SamWilsn) [reply notion-server/src/ethereum.rs:31]: > allocating / creating a new object seems like what I'd expect to happen in a From statement like this

Yeah, allocating the internal struct will happen either way. I'm more concerned about the `String`s and `Vec`s inside.

> Seems like making it force moves would be a more restrictive API

If you pass-by-move, you give the caller the choice between making copies or not. If you pass by reference, you always make a copy, even if you don't need to.

Here's how pass-by-move gives the caller the choice:

```rust
fn pass_by_move_cpy() {
    let obj = SomeBigObject::new();
    led jbo = OtherBigObject::from(obj.clone()); // The caller decides to make a copy.

    // The caller needs obj here, for whatever reason.
    println!("{:?}", obj);
}

fn pass_by_move_nocpy() {
    let obj = SomeBigObject::new();
    led jbo = OtherBigObject::from(obj);
    // The caller doesn't need obj any longer, so moving it save allocations.
}
```

If you use a `From<&Foo>`, it isn't possible to implement `pass_by_move_nocpy`.

This entire discussion is pretty much moot here, since the code has to convert from a `String` to a `Vec` anyway. Making this change only makes sense if you apply [this suggestion](https://github.com/quilt/simulation/pull/14#discussion_r350313031).

Comment (SamWilsn) [reply notion-server/src/ethereum.rs:31]: > premature optimization

Absolutely! But it's a good opportunity to discuss API design in Rust ;3

File: notion-server/src/ethereum.rs (hunk 5 above)
Comment (SamWilsn) [reply notion-server/src/ethereum.rs:61]: I think whatever arguments apply for using `EeIndex` internally also apply externally.

These kind of types are called `newtypes` and are fairly useful to organize code, and prevent accidentally passing a shard index into a method expecting an ee index, for example.

File: notion-server/src/ethereum.rs (hunk 7 above)
Comment (SamWilsn) [reply notion-server/src/ethereum.rs:19]: `notion_server::ethereum::ShardBlock` will never be visible outside of the `ethereum` module, so outside users won't be confused. They won't even be aware there are two similarly named structs.

The only code which has to refer to both `args::ShardBlock` and `ShardBlock` lives under the `ethereum` module, so I think that'll be plenty clear.

If it gets particularly confusing, we can always rename the module as we go.

Comment (gjtrowbridge) [reply notion-server/src/ethereum.rs:19]: Alright, I still prefer interface_args, but I don't feel too strongly, so I'm happy to swap back to `args`

File: notion-server/src/ethereum.rs (hunk 2 above)
Comment (gjtrowbridge) [reply notion-server/src/ethereum.rs:28]: If you're ok with it I may leave this change until after we set up the web server, that way it'll be easier to port over the existing encode/decode stuff.

One other thought that occurred to me (and the reason I think I set it up this way in the first place) is that in this implementation we can have the web server accept this struct directly as JSON, removing the need for a 3rd intermediate struct.

File: notion-server/src/ethereum.rs (hunk 4 above)
Comment (gjtrowbridge) [reply notion-server/src/ethereum.rs:31]: I think for now it will be simpler to keep this as-is with references, as it makes a number of things easier -- no need to clone internal structs, it allows the use of the `map` function on `vec`, etc. If we feel strongly about changing this later, we could do that then.  Also, at least for the return values (converting from internal to interface structs), we'd pretty much always want to re-allocate, since we don't want to change ownership of the internal structs.

So, unless you feel strongly, I'd prefer to leave it as-is for now, but if you feel strongly let me know and I'm cool with updating it.

I do agree that having it be `move` by default leaves the decision to `clone` to the caller of the function. This is a small distinction, but worth noting that there could also be some cases where the result of the `from` function returns a subset of the fields of the original struct, that could be less efficient because the `clone` would re-allocate all fields instead of just the ones necessary.

File: notion-server/src/ethereum.rs (hunk 2 above)
Comment (SamWilsn) [reply notion-server/src/ethereum.rs:28]: > removing the need for a 3rd intermediate struct.

I'm not suggesting a third intermediate struct. I'm suggesting deriving `Deserialize` directly on this struct, and using serde's `with` annotation.

I'm okay with leaving it for now though.

File: notion-server/src/ethereum.rs (hunk 4 above)
Comment (SamWilsn) [reply notion-server/src/ethereum.rs:31]: > it allows the use of the map function on vec

You're looking for `into_iter` I think.

The rest are great points. I'm fine either way :3

**Review by adietrichs — APPROVED** [review]

**Comment by gjtrowbridge** [comment]
These changes merged via another PR.
